ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/Iterator.hs

```diff
@@ -257,24 +258,49 @@ newIterator itEnv@IteratorEnv{..} getItEnv registry from to = do
           => ExceptT (UnknownRange blk) m (DeserialisableIterator m blk)
     start = lift (ImmDB.getTipInfo itImmDB) >>= \case
       Origin                          -> findPathInVolDB
-      At (tipSlot, tipHash, _tipIsEBB) ->
+      At (tipSlot, tipHash, tipIsEBB) ->
         case atSlot endPoint `compare` tipSlot of
           -- The end point is < the tip of the ImmutableDB
           LT -> streamFromImmDB
           EQ | withHash endPoint == tipHash
                 -- The end point == the tip of the ImmutableDB
              -> streamFromImmDB
-             | otherwise
-                -- The end point /= the tip of the ImmutableDB. Either the
-                -- fork is too old, or the end point (or the tip of the
-                -- ImmutableDB) is an EBB. For example, the tip of the
-                -- ImmutableDB is an EBB and the end point is the regular
-                -- block after it.
+
+             -- The end point /= the tip of the ImmutableDB.
+             --
+             -- There are three possibilities:
+             --
+             -- 1. The end point is the regular block (or a different EBB?)
+             --    with the same slot number as the EBB which is currently at
+             --    the tip of the ImmutableDB. In this case, we must look in
+             --    the VolatileDB for the block, as we know it is not in the
+             --    ImmutableDB.
+             --
+             -- 2. The end point is the EBB with the same slot number as the
+             --    regular block which is currently at the tip of the
+             --    ImmutableDB. In this case, we can stream from just the
+             --    ImmutableDB, as the end point is in the ImmutableDB.
+             --
+             -- 3. The fork is too old, i.e., the end point refers to a block
+             --    (or EBB) not on the current chain, but with the same slot
+             --    number as our tip.
+             --
+             -- We don't know upfront with which of the three cases we're
+             -- dealing, as we don't know which block the end point refers to
+             -- without trying to read it from the ImmutableDB or VolatileDB.
+             -- The important thing is to also try the ImmutableDB in case 2.
+             | IsEBB <- tipIsEBB  -- Case 1 or 3
              -> findPathInVolDB
+             | otherwise  -- Case 2 or 3
+             -> streamFromImmDB `fallbackTo` findPathInVolDB
```

Comment:
Related to my comment on Case 3 above: when do we need to fall back onto `findPathInVolDB` here?

We reach this alternative when the end point and the tip have the same slot, they also have differing hashes, and the tip is not an EBB. So isn't Case 2 (covered by `streamFromImmDB`) the only possibility that should _not_ raise a `ForkTooOld` exception?

Reply:
👍

Reply:
New code looks great.